File: March 2023/Day 24/Remove and Reverse/remove-and-reverse.java
//{ Driver Code Starts
//Initial Template for Java
import java.io.*;
import java.util.*; 
class GFG{
    public static void main(String args[]) throws IOException { 
        BufferedReader read = new BufferedReader(new InputStreamReader(System.in));
        int t = Integer.parseInt(read.readLine());
        
        while(t-- > 0){
          
            String S = read.readLine().trim();
            Solution ob = new Solution();
            String ans = ob.removeReverse(S); 
            System.out.println(ans);
        }
    } 
} 

// } Driver Code Ends
//User function Template for Java

class Solution 
{ 
    String removeReverse(String S) 
    { 
        // code here
        int[] fre = new int[26];
        for(int i=0;i<S.length();i++) fre[S.charAt(i)-'a']++;
        int i=0,j=S.length()-1,time=0;
        byte[] isRemoved = new byte[S.length()];
        while(i<=j) {
            if(time%2==0) {
                if(fre[S.charAt(i)-'a']>1) {
                    time++;
                    isRemoved[i]=1;
                    fre[S.charAt(i)-'a']--;
                }
                i++;
            } else {
                if(fre[S.charAt(j)-'a']>1) {
                    time++;
                    isRemoved[j]=1;
                    fre[S.charAt(j)-'a']--;
                }
                j--;
            }
        }
        StringBuilder sb = new StringBuilder();
        for(i=0;i<S.length();i++) if(isRemoved[i]==0) sb.append(S.substring(i,i+1));
        if(time%2==1) sb.reverse();
        return sb.toString();
    }
} 

//{ Driver Code Starts.

// } Driver Code Ends

File: March 2023/Day 24/Remove and Reverse/remove-and-reverse.cpp
//{ Driver Code Starts
//Initial Template for C++
#include <bits/stdc++.h>
using namespace std;

// } Driver Code Ends
//User function Template for C++
class Solution {
  public:
    string removeReverse(string S) {
        // code here
        vector<int> fre(26,0);
        for(int i=0;i<S.length();i++) fre[S[i]-'a']++;
        vector<int> isRemoved(S.length(),0);
        int i=0,j=S.length()-1,time=0;
        while(i<=j) {
            if(time%2==0) {
                if(fre[S[i]-'a']>1) {
                    time++;
                    isRemoved[i]=1;
                    fre[S[i]-'a']--;
                }
                i++;
            } else {
                if(fre[S[j]-'a']>1) {
                    time++;
                    isRemoved[j]=1;
                    fre[S[j]-'a']--;
                }
                j--;
            }
        }
        string sb="";
        for(i=0;i<S.length();i++) if(isRemoved[i]==0) sb+=S[i];
        if(time%2==1) reverse(sb.begin(),sb.end());
        return sb;
    }
};

//{ Driver Code Starts.
int main() {
    int t;
    cin >> t;
    while (t--) {
        string S;
        cin>>S;
        Solution obj;
        auto ans = obj.removeReverse(S);
        cout<<ans<<endl;
    }
}
// } Driver Code Ends